plugins/fpc/fu-fpc-device.c

```diff
@@ -0,0 +1,521 @@
+		g_debug("already in bootloader mode, skipping");
+		return TRUE;
+	}
+	if (!fu_fpc_device_fw_cmd(self, FPC_CMD_BOOT0, NULL, 0, FALSE, error))
+		return FALSE;
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_setup(FuDevice *device, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* setup */
+	if (!FU_DEVICE_CLASS(fu_fpc_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	self->is_moc_device = (fu_usb_device_get_pid(FU_USB_DEVICE(device)) >> 2 & 0x1) ? TRUE
+											: FALSE;
+
+	if (!fu_fpc_device_setup_mode(self, error)) {
+		g_prefix_error(error, "failed to get device mode: ");
+		return FALSE;
+	}
+
+	/* ensure version */
+	if (!fu_fpc_device_setup_version(self, error)) {
+		g_prefix_error(error, "failed to get firmware version: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_write_firmware(FuDevice *device,
+			     FuFirmware *firmware,
+			     FuProgress *progress,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+	FuFpcDfuStatus dfu_status = {0};
+	gboolean ret_dnload = TRUE;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, "init");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 99, NULL);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* don't auto-boot firmware */
+	if (!fu_fpc_device_update_init(self, &error_local)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "failed to initial update: %s",
+			    error_local->message);
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* build packets */
+	chunks = fu_chunk_array_new_from_bytes(fw,
+					       0x00,
+					       0x00, /* page_sz */
+					       self->max_block_size);
+
+	/* write each block */
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) req = g_byte_array_new();
+
+		g_byte_array_append(req, fu_chunk_get_data(chk), fu_chunk_get_data_sz(chk));
+
+		if (!fu_fpc_device_dfu_cmd(self,
+					   FPC_CMD_DFU_DNLOAD,
+					   (guint16)i,
+					   req->data,
+					   (gsize)req->len,
+					   FALSE,
+					   FALSE,
+					   &error_local)) {
+			ret_dnload = FALSE;
+			break;
+		}
+
+		if (!fu_device_retry_full(device,
+					  fu_fpc_device_check_dfu_status,
+					  FPC_DFU_MAX_ATTEMPTS,
+					  20,
+					  (gpointer)&dfu_status,
+					  &error_local)) {
+			ret_dnload = FALSE;
+			break;
+		}
+
+		/* update progress */
+		fu_progress_set_percentage_full(fu_progress_get_child(progress),
+						(gsize)i + 1,
+						(gsize)chunks->len);
+	}
+
```

Comment:
i think you probably want to `fu_progress_step_done(progress);` here and add another `fu_progress_add_step` for the checking DFU status.